packages/cli/demo/cat.js

```diff
@@ -585,6 +644,138 @@ const chatComponent = ($parent, powers, { dismissChat }) => {
+  let $error = $parent.querySelector('#eval-error');
+  const $endowments = new Map();
+
+  const clearEval = () => {
+    $source.value = '';
+    $resultName.value = '';
+    for (const $endowment of $endowments.values()) {
+      $endowment.remove();
+    }
+    $endowments.clear();
+    const $newError = document.createTextNode('');
+    $error.replaceWith($newError);
+    $error = $newError;
+    dismissEval();
+  };
+
+  $discard.addEventListener('click', () => {
+    clearEval();
+  });
+
+  const handleRemoveEndowment = event => {
+    const { target } = event;
+    const id = target.getAttribute('id');
+    const $endowment = $endowments.get(id);
+    if ($endowment === undefined) {
+      throw new Error(`Endowment does not exist for id ${id}`);
+    }
+    $endowments.delete(id);
+    $endowment.remove();
+  };
+
+  const handleAddEndowment = () => {
+    const $endowment = document.createElement('p');
+    const codeNameId = nextId();
+    const petNameId = nextId();
+    const removeId = nextId();
+
+    $endowment.innerHTML = `
+      <label for="${codeNameId}">
+        Name in source:
+        <input id="${codeNameId}" type="text" class="big half-wide code-name">
+      </label>
+      <label for="${petNameId}">
+        Pet name:
+        <input id="${petNameId}" type="text" class="big half-wide pet-name">
+      </label>
+      <button id="${removeId}">Remove</button>
+    `;
+
+    const $remove = $endowment.querySelector(`#${removeId}`);
+    $remove.addEventListener('click', handleRemoveEndowment);
+    $endOfEndowments.parentElement.insertBefore($endowment, $endOfEndowments);
+    $endowments.set(removeId, $endowment);
+
+    const $codeName = $endowment.querySelector('.code-name');
+    $codeName.focus();
+  };
+
+  $addEndowment.addEventListener('click', handleAddEndowment);
+
+  const handleEval = () => {
+    const source = $source.value;
+    const workerName = 'MAIN';
+    const names = Array.from($endowments.values(), $endowment => {
+      const $codeName = $endowment.querySelector('.code-name');
+      const $petName = $endowment.querySelector('.pet-name');
+      return {
+        petName: $petName.value,
+        codeName: $codeName.value,
+      };
+    });
+    const codeNames = names.map(({ codeName }) => codeName);
+    const petNames = names.map(({ petName }) => petName);
+    const resultName = $resultName.value.trim() || undefined;
+    E(powers)
+      .evaluate(workerName, source, codeNames, petNames, resultName)
+      .then(clearEval, error => {
+        const $newError = document.createElement('p');
+        $newError.className = 'error';
+        $newError.innerText = error.message;
+        $error.replaceWith($newError);
+        $error = $newError;
+      });
+  };
+
+  $frame.addEventListener('keyup', event => {
+    const { key, repeat, metaKey } = event;
+    if (repeat || metaKey) return;
+    if (key === 'Enter') {
+      event.stopPropagation();
+      handleEval();
+    }
```

Comment:
This triggers `handleEval` too eagerly, namely whenever enter is pressed while the eval window is open. This causes errors when, for example, adding newlines in the source input or attempting to add an endowment using only the keyboard. IMO, form submission should only trigger on enter if the submission button is selected.

It's not part of this PR, but we should perhaps review `handleAdopt` for the same issue.

Reply:
Tough call whether to oblige you to use shift+enter or navigate over to the submit button, but I’ve done the latter.